chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2479 @@
+    "save": SAVE_LIST,
+    "scode": PTTIDCODE_LIST,
+    "screv": RESUME_LIST,
+    "sftd": SHIFTD_LIST,
+    "stedelay": STEDELAY_LIST,
+    "step": STEP_LIST,
+    "step291": STEP291_LIST,
+    "tdrab": TDRAB_LIST,
+    "tdrch": TDRCH_LIST,
+    "timeout": TIMEOUT_LIST,
+    "txled": COLOR_LIST,
+    "txpower": TXPOWER_LIST,
+    "txpower3": TXPOWER3_LIST,
+    "voice": VOICE_LIST,
+    "vox": VOX_LIST,
+    "widenarr": BANDWIDTH_LIST,
+    "workmode": WORKMODE_LIST,
+    "wtled": COLOR_LIST
+}
+
+GMRS_FREQS1 = [462.5625, 462.5875, 462.6125, 462.6375, 462.6625,
+               462.6875, 462.7125]
+GMRS_FREQS2 = [467.5625, 467.5875, 467.6125, 467.6375, 467.6625,
+               467.6875, 467.7125]
+GMRS_FREQS3 = [462.5500, 462.5750, 462.6000, 462.6250, 462.6500,
+               462.6750, 462.7000, 462.7250]
+GMRS_FREQS_B = GMRS_FREQS2 * 3 + GMRS_FREQS3 * 2
+
+GMRS_FREQ_A = GMRS_FREQS1 + GMRS_FREQS3
+
+HAM_FREQS = [162.55000, 162.40000, 162.47500, 162.42500, 162.45000,
+             162.50000, 162.52500, 161.65000, 161.77500, 161.75000]
+HAM_NAME = ["NOAA 1", "NOAA 2", "NOAA 3", "NOAA 4", "NOAA 5", "NOAA 6",
+            "NOAA 7", "NOAA 8", "NOAA 9", "NOAA 10", "NOAA 11"]
+
+ALL_MODEL = ["TD-H8", "TD-H8-HAM", "TD-H8-GMRS"]
+
+
+def _do_status(radio, block):
+    status = chirp_common.Status()
+    status.msg = "Cloning"
+    status.cur = block
+    status.max = radio._memsize
+    radio.status_fn(status)
+
+
+TD_H8 = b"\x50\x56\x4F\x4A\x48\x1C\x14"
+
+
+def _upper_band_from_data(data):
+    return data[0x03:0x04]
+
+
+def _upper_band_from_image(radio):
+    return _upper_band_from_data(radio.get_mmap())
+
+
+def _firmware_version_from_data(data, version_start, version_stop):
+    version_tag = data[version_start:version_stop]
+    return version_tag
+
+
+def _firmware_version_from_image(radio):
+    version = _firmware_version_from_data(radio.get_mmap(),
+                                          radio._fw_ver_file_start,
+                                          radio._fw_ver_file_stop)
+    # LOG.debug("_firmware_version_from_image: " + util.hexprint(version))
+    return version
+
+
+def _do_ident(radio, magic, secondack=True):
+    serial = radio.pipe
+    serial.timeout = 1
+
+    LOG.info("Sending Magic: %s" % util.hexprint(magic))
+    serial.write(magic)
+    ack = serial.read(1)
+
+    if ack != b"\x06":
+        if ack:
+            pass
+            # LOG.debug(repr(ack))
+        raise errors.RadioError("Radio did not respond")
+
+    serial.write(b"\x02")
+
+    response = b""
+    for i in range(1, 9):
+        byte = serial.read(1)
+        response += byte
+        if byte == b"\xDD":
+            break
+
+    if len(response) in [8, 12]:
+        # DEBUG
+        LOG.info("Valid response, got this:")
+        LOG.info(util.hexprint(response))
+
+        # Determine the walkie-talkie mode
+        # TDH8 have three mode:ham, gmrs and normal
+        global MODE_RADIO
+        MODE_RADIO = util.hexprint(response)[31:37]
```

Comment:
As I mentioned before, you may not use a global for this. Also using the output of hexprint is not appropriate for inspection, it's for debugging.

Reply:
I deleted the output using hexprint, so I processed the data in L695 do_download() to get the walkie-talkie ID, but it is still a global variable, because I need to check the walkie-talkie ID in L922 sync_in(), but I don't want to re-execute L695 do_download( ) This will re-execute the download, so I use a global variable, how should I deal with this walkie-talkie identification variable?

Reply:
Set a flag on the `radio` object. The `radio` is created per-use and lives for the lifecycle of one conversation/edit session with a given model. So make it so that once the radio talks to a specific model a flag gets set and then apply the rules for that model when editing, and again when communicating. You also need to be able to detect that flag when you load an image so that you know which model a given image file is for, so you can agree or refuse to talk to that model if the user opens an image from disk.

Reply:
In the latest version, I don't use the global.